### PR TITLE
fix(backend): coerce tz-naive action-item dates to UTC (#11137)

### LINKED ISSUE #11137: Bug: tz-naive action-item dates cause Firestore reject / silent task loss
## Summary

`_prepare_action_item_for_write` in `backend/database/action_items.py` only normalizes **string** date fields (`fromisoformat` + drop on `ValueError`). It does **not** coerce:

- tz-naive `datetime` objects
- date-only ISO strings like `"2024-01-01"` (which `fromisoformat` parses as naive)

Firestore rejects timezone-naive datetimes. When conversation postprocess saves extracted action items via `create_action_items_batch` on the fire-and-forget `postprocess_executor` path (`utils/conversations/process_conversation.py` → `_save_action_items`), a failed batch can mean **extracted tasks are silently lost**.

## Why this happens

- REST API models use `AwareDatetime` for dates.
- LLM / structured extraction (`models/structured_extraction.py` `ExtractedActionItem.due_at`, `models/structured.py` `ActionItem.due_at`) uses plain `Optional[datetime]`, so due dates often arrive naive or as date-only strings.

Sibling coerce patterns already exist elsewhere:

- `backend/utils/mcp_action_items.py` `parse_due_at`
- `backend/database/api_key_metadata.py` `_coerce_utc_datetime`

## Expected

Write preparation attaches UTC to naive datetimes (and normalizes aware values to UTC) before Firestore writes, matching the MCP / API-key helpers. Malformed strings still drop with a warning (no 500).

## Suggested fix

In `_prepare_action_item_for_write`, after parsing strings (and for existing `datetime` values): if `tzinfo is None` → `.replace(tzinfo=timezone.utc)`. Extend `backend/tests/unit/test_action_items_date_coercion.py` (naive datetime, date-only string, aware → UTC).

## Impact

High — silent loss of user-visible action items after conversation processing when the model emits a due date without timezone.

### PR DESCRIPTION
**Pull request by aryanorastar**
## What changed and why

Coerce timezone-naive action-item date fields to UTC in `_prepare_action_item_for_write` so Firestore batch creates cannot reject LLM/extracted `due_at` values and silently drop tasks on the fire-and-forget postprocess path. Fixes #11137.

Matches sibling helpers `_coerce_utc_datetime` (`database/api_key_metadata.py`) and `parse_due_at` (`mcp_action_items.py`): naive/`YYYY-MM-DD` → UTC, malformed dropped. **Aware values are also normalized to UTC** via `astimezone(utc)` (same as those siblings); previously aware datetimes passed through with their offset preserved — a no-op for Firestore storage, but called out so reviewers aren't surprised. Empty-string `''` is now popped rather than left in the dict.

## Product invariants affected

none

## How it was verified

```bash
cd backend && python3 -m pytest tests/unit/test_action_items_date_coercion.py -q
# 6 passed
```

Confirmed extraction models use plain `Optional[datetime]` (`structured_extraction.ExtractedActionItem.due_at`) while Firestore requires aware datetimes; date-only ISO strings parse naive via `fromisoformat`.

Unit tests verify **coercion output** (the bug surface); the Firestore SDK itself is not exercised here — same shape as the sibling helper tests.

## Tests

`backend/tests/unit/test_action_items_date_coercion.py` — naive datetime → UTC, date-only ISO → UTC, aware offset → UTC, malformed string still dropped, one bad field doesn't kill siblings.

## Failure class (fixes)

Failure-Class: none

## Scoped cleanups (optional)

n/a

### CODE REVIEW

**Comment by aryanorastar** [comment]
Updated the PR body with two reviewer-facing clarifications (no code change):

1. Aware datetimes are also normalized to UTC (matches `_coerce_utc_datetime` / `parse_due_at`; Firestore stores UTC anyway).
2. Unit coverage asserts coercion output; Firestore SDK write path is not exercised in these tests.

**Review by Git-on-my-level — DISMISSED** [review]
Thanks for the focused fix and tests. I’m marking one blocker before this is merge-ready: the new tolerant date handling still only catches `ValueError` around `fromisoformat()`, while the later UTC normalization can raise `OverflowError` for boundary aware values, e.g. an otherwise parseable datetime whose offset conversion moves it outside Python’s datetime range.

Because this helper is meant to prevent a single bad action-item date from failing create/update or a batch write, please wrap the parse + UTC normalization path so `OverflowError` is handled the same way as malformed strings (drop that field and keep processing siblings), and add a regression test for that boundary case.

---
*by AI on behalf of David — if you need David’s attention urgently, please @Git-on-my-level and escalate with `need human response`.*

File: backend/database/action_items.py [hunk 1]
```diff
@@ -127,19 +127,44 @@ def _prepare_action_item_for_write(action_item_data: Dict[str, Any], *, partial:
...
+                date_field,
+                type(value).__name__,
+            )
+            action_item_data.pop(date_field, None)
+            continue
+
+        if parsed.tzinfo is None:
+            parsed = parsed.replace(tzinfo=timezone.utc)
+        else:
+            parsed = parsed.astimezone(timezone.utc)
```
[the remark this answers is not in the document]
Comment (aryanorastar) [reply backend/database/action_items.py]: Addressed in 73f2b829.

Parse + UTC normalization now share one `except (OverflowError, ValueError)` path (same shape as `api_key_metadata._coerce_utc_datetime`). Boundary aware values that overflow on `astimezone(UTC)` drop that field and keep sibling fields — covered by `test_overflow_on_utc_normalization_is_dropped_not_raised` (+ ISO-string variant).

**Comment by aryanorastar** [comment]
## Addressed review @73f2b829

@Git-on-my-level Blocker fixed:

- Parse + UTC normalize wrapped in `except (OverflowError, ValueError)` — same drop-field behavior as malformed strings.
- Regression tests: `datetime(1,1,1,+14h)` and `0001-01-01T00:00:00+14:00` drop `due_at` without raising; sibling `created_at` still coerces.

```bash
cd backend && python3 -m pytest tests/unit/test_action_items_date_coercion.py -q
# 8 passed
```

**Comment by aryanorastar** [comment]
@Git-on-my-level need human response — re-review request.

Your OverflowError blocker is fixed on head `@73f2b829` (`except (OverflowError, ValueError)` around parse+UTC normalize, with boundary tests). CI green. Could you dismiss the stale CHANGES_REQUESTED / APPROVE?

**Review by Git-on-my-level — COMMENTED** [review]
Thanks for the focused follow-up. I re-checked the current head and the previous blocker looks resolved: `_prepare_action_item_for_write` now catches `OverflowError` across the parse/UTC-normalization path, drops only the bad date field, and preserves sibling fields. The added boundary tests cover both an aware `datetime` and an ISO string whose offset conversion would underflow.

I also ran the focused backend unit test locally in an isolated venv:

```bash
python -m pytest tests/unit/test_action_items_date_coercion.py -q
# 8 passed
```

CI is green as well. I’m not formally approving because this touches backend persistence/date normalization and is tied to the #11137 fix path, so it should still get maintainer eyes before merge, but from this automation review the requested change is addressed.

---
*by AI on behalf of David — if you need David’s attention urgently, please @Git-on-my-level and escalate with `need human response`.*

**Comment by aryanorastar** [comment]
Author-side clear for merge.

@Git-on-my-level dismissed the OverflowError CR on `@73f2b829` (parse+UTC normalize drops overflow dates; 8 unit tests; CI green; MERGEABLE). Technical read positive — still needs **write-access APPROVE**.

@kodjima33 — when you have a moment, please APPROVE/merge (tz-naive action-item dates → UTC so Firestore batch creates don’t silently drop tasks, #11137). Shrink-only backend persistence helper; no product invariant change.

**Comment by aryanorastar** [comment]
@kodjima33 final approval reminder: PR #11138 is fully CI-green at 73f2b82948. David dismissed the OverflowError changes-request after verifying the boundary handling and 8/8 focused tests; the fix remains a small Firestore date-normalization guard with no product-invariant change. The only remaining gate is write-access approval/merge.

**Comment by aryanorastar** [comment]
@kodjima33 merge ping — still blocked only on write-access APPROVE.

Head `@73f2b829` · MERGEABLE · CI green (backend unit + hermetic gates) · David OverflowError CR dismissed · `positive-signal` · shrink-only `_prepare_action_item_for_write` UTC coerce for #11137 (stops Firestore batch creates from silently dropping tasks).

Please APPROVE/merge when you have a moment.

**Comment by aryanorastar** [comment]
@kodjima33 **need human response** — write-access APPROVE / merge is the only remaining blocker on #11138.

Author-clear on tip `1fab964ceb3f` (refreshed onto current main; was ~463 behind):

- David OverflowError CR **dismissed** · `positive-signal`
- Local re-check: `test_action_items_date_coercion.py` **8 passed**
- Shrink-only `_prepare_action_item_for_write` UTC coerce for #11137 (stops Firestore batch creates from silently dropping tasks)
- Failure-Class: none · no product invariant change

CI re-running on the main merge. Please APPROVE/merge when you can.

**Comment by Git-on-my-level** [comment]
Re-checked the refreshed head `1fab964ceb3f6b2eca894f60165b9ef54995d048`; the prior OverflowError blocker remains resolved.

What I verified:
- `backend/database/action_items.py`: `_prepare_action_item_for_write` now normalizes all four date fields to timezone-aware UTC, handles strings and `datetime` objects, drops empty/non-datetime/bad date values field-by-field, and wraps both parsing plus UTC conversion in `except (OverflowError, ValueError)`. That means boundary offset values should no longer fail the whole action-item write or batch while sibling fields continue to be preserved.
- `backend/tests/unit/test_action_items_date_coercion.py`: coverage now includes malformed strings, date-only ISO strings, tz-naive datetimes, aware-offset normalization, sibling preservation, and both aware-`datetime` and ISO-string overflow cases.

I also ran the focused backend unit test in a clean local venv on this head:

```bash
python -m pytest tests/unit/test_action_items_date_coercion.py -q
# 8 passed, 1 warning
```

No further changes requested from this automation review. Keeping this as a positive signal rather than a formal approval because this is a backend persistence/date-normalization fix path and still needs write-access maintainer sign-off before merge.

---
*by AI on behalf of David — backend persistence/date-normalization change needs maintainer sign-off before merge.*

**Comment by aryanorastar** [comment]
@kodjima33 need human response — write-access APPROVE / merge for #11138 / #11137.

Author-clear on `1fab964ceb`:

- David OverflowError CR dismissed; `positive-signal`; no further code changes
- **MERGEABLE** + all checks green; only branch-protection write-access APPROVE remaining
- Shrink-only: `_prepare_action_item_for_write` coerces naive/ISO dates to UTC and drops overflow/malformed fields so Firestore batch creates don’t silently lose tasks
- Local: `test_action_items_date_coercion.py` → 8 passed · Failure-Class: none

Please APPROVE/merge when you can.

**Review by kodjima33 — APPROVED** [review]
5/5: root cause (Firestore rejects tz-naive datetimes on the fire-and-forget postprocess path), fixes #11137, scoped, CI green, regression tests added. Mirrors _coerce_utc_datetime / parse_due_at.
